# Data Types and Formatting

## INFO 442 Group 5 Shuyang Zhang 320230942711

In [4]:
import pandas as pd
import numpy as np
mpg = pd.read_csv('mpg.csv')

### 1. Load the seaborn mpg dataset. Identify and fix all type issues (hint: check horsepower).

In [18]:
print("Original dtypes:")
print(mpg.dtypes)
print()

print("horsepower unique values (first 10):")
print(mpg['horsepower'].unique()[:10])
print()

mpg['horsepower'] = pd.to_numeric(mpg['horsepower'], errors='coerce')
print("After conversion - dtypes:")
print(mpg.dtypes)
print(f"Missing horsepower after fix: {mpg['horsepower'].isnull().sum()}")
print()


Original dtypes:
mpg                        float64
cylinders                    int64
displacement               float64
horsepower                 float64
weight                       int64
acceleration               float64
model_year                   int64
origin                      object
name                        object
model_date          datetime64[ns]
manufacturer                object
manufacturer_obj            object
manufacturer_cat          category
name_normalized             object
dtype: object

horsepower unique values (first 10):
[130. 165. 150. 140. 198. 220. 215. 225. 190. 170.]

After conversion - dtypes:
mpg                        float64
cylinders                    int64
displacement               float64
horsepower                 float64
weight                       int64
acceleration               float64
model_year                   int64
origin                      object
name                        object
model_date          datetime64[ns]
manufacture

answer: Horsepower was the only column with type issues. After conversion, it became float64 with 6 missing values.

### 2. The mpg dataset has a model_year column stored as integers (e.g. 70 = 1970). Convert it to a proper datetime (use January 1 of that year). 

In [23]:
mpg['model_date'] = pd.to_datetime(mpg['model_year'].astype(str) + '-01-01', format='%y-%m-%d')
print("Model year as datetime:")
print(mpg[['model_year', 'model_date']].head())
print(f"Date range: {mpg['model_date'].min()} to {mpg['model_date'].max()}")
print()

Model year as datetime:
   model_year model_date
0          70 1970-01-01
1          70 1970-01-01
2          70 1970-01-01
3          70 1970-01-01
4          70 1970-01-01
Date range: 1970-01-01 00:00:00 to 1982-01-01 00:00:00



### 3. The name column contains strings like 'chevrolet chevelle malibu'. Extract the manufacturer (first word) into a new column and convert it to category dtype. How much memory does this save?

In [26]:
mpg['manufacturer'] = mpg['name'].str.split().str[0]

mpg['manufacturer_obj'] = mpg['manufacturer'].astype('object')
mpg['manufacturer_cat'] = mpg['manufacturer'].astype('category')

obj_mem = mpg['manufacturer_obj'].memory_usage(deep=True)
cat_mem = mpg['manufacturer_cat'].memory_usage(deep=True)

print(f"Object dtype memory: {obj_mem / 1024:.1f} KB")
print(f"Category dtype memory: {cat_mem / 1024:.1f} KB")
print(f"Memory reduction: {(1 - cat_mem/obj_mem)*100:.1f}%")
print()

print(f"Number of unique manufacturers: {mpg['manufacturer'].nunique()}")
print(f"Sample manufacturers: {mpg['manufacturer'].unique()[:10]}")
print()


Object dtype memory: 21.5 KB
Category dtype memory: 3.6 KB
Memory reduction: 83.5%

Number of unique manufacturers: 37
Sample manufacturers: ['chevrolet' 'buick' 'plymouth' 'amc' 'ford' 'pontiac' 'dodge' 'toyota'
 'datsun' 'volkswagen']



answer: Converting the manufacturer column from object to category dtype saved 17.9 KB of memory (from 21.5 KB to 3.6 KB), a reduction of 83.5%.

### 4. Write a function normalize_string_column(series) that applies strip, lower, and collapses internal spaces, then apply it to the name column.

In [3]:
def normalize_string_column(series):
    return (series
        .astype(str)
        .str.strip()
        .str.lower()
        .str.replace(r'\s+', ' ', regex=True)
    )

mpg['name_normalized'] = normalize_string_column(mpg['name'])

